data/transactions/logic/assembler.go

```diff
@@ -1016,58 +1038,74 @@ func int8Imm(value string, label string) (byte, error) {
 	return byte(res), err
 }
 
-func asmItxn(ops *OpStream, spec *OpSpec, args []string) error {
+func asmItxn(ops *OpStream, spec *OpSpec, mnemonic token, args []token) *sourceError {
 	if len(args) == 1 {
-		return asmDefault(ops, spec, args)
+		return asmDefault(ops, spec, mnemonic, args)
 	}
 	if len(args) == 2 {
 		itxna := OpsByName[ops.Version]["itxna"]
-		return asmDefault(ops, &itxna, args)
+		return asmDefault(ops, &itxna, mnemonic, args)
 	}
-	return ops.errorf("%s expects 1 or 2 immediate arguments", spec.Name)
+	return mnemonic.errorf("%s expects 1 or 2 immediate arguments", spec.Name)
 }
 
 // asmGitxn substitutes gitna's spec if the are 3 args
-func asmGitxn(ops *OpStream, spec *OpSpec, args []string) error {
+func asmGitxn(ops *OpStream, spec *OpSpec, mnemonic token, args []token) *sourceError {
 	if len(args) == 2 {
-		return asmDefault(ops, spec, args)
+		return asmDefault(ops, spec, mnemonic, args)
 	}
 	if len(args) == 3 {
 		itxna := OpsByName[ops.Version]["gitxna"]
-		return asmDefault(ops, &itxna, args)
+		return asmDefault(ops, &itxna, mnemonic, args)
 	}
-	return ops.errorf("%s expects 2 or 3 immediate arguments", spec.Name)
+	return mnemonic.errorf("%s expects 2 or 3 immediate arguments", spec.Name)
 }
 
-func asmItxnField(ops *OpStream, spec *OpSpec, args []string) error {
-	if len(args) != 1 {
-		return ops.errorf("%s expects one argument", spec.Name)
+func asmItxnField(ops *OpStream, spec *OpSpec, mnemonic token, args []token) *sourceError {
+	if err := ops.checkArgCount(spec.Name, mnemonic, args, 1); err != nil {
+		return err
 	}
-	fs, ok := txnFieldSpecByName[args[0]]
+	fs, ok := txnFieldSpecByName[args[0].str]
 	if !ok {
-		return ops.errorf("%s unknown field: %#v", spec.Name, args[0])
+		return args[0].errorf("%s unknown field: %#v", spec.Name, args[0].str)
 	}
 	if fs.itxVersion == 0 {
-		return ops.errorf("%s %#v is not allowed.", spec.Name, args[0])
+		return args[0].errorf("%s %#v is not allowed.", spec.Name, args[0].str)
 	}
 	if fs.itxVersion > ops.Version {
-		return ops.errorf("%s %s field was introduced in v%d. Missed #pragma version?", spec.Name, args[0], fs.itxVersion)
+		return args[0].errorf("%s %s field was introduced in v%d. Missed #pragma version?", spec.Name, args[0].str, fs.itxVersion)
 	}
 	ops.pending.WriteByte(spec.Opcode)
 	ops.pending.WriteByte(fs.Field())
 	return nil
 }
 
-type asmFunc func(*OpStream, *OpSpec, []string) error
+type asmFunc func(*OpStream, *OpSpec, token, []token) *sourceError
 
-// Basic assembly. Any extra bytes of opcode are encoded as byte immediates.
-func asmDefault(ops *OpStream, spec *OpSpec, args []string) error {
-	expected := len(spec.OpDetails.Immediates)
-	if len(args) != expected {
+func (ops *OpStream) checkArgCount(name string, mnemonic token, args []token, expected int) *sourceError {
```

Comment:
jj, could you please explain this function?

It looks like handling immediate argument count, but this is too dense, and it is called in roughly 10 places. (so I assume about half of the changes in this file are related with new check)

Reply:
The function catching my interest is that, a bunch of `asm*` methods are extending one more field `mnemonic token`, while I haven't wrapped my head around for the need.

Reply:
Almost all of that code is there in order to get just the right `error` value.  The basic function is just:
```
	offered := len(args)
	if offered != expected {
           // return a good error
        }
        return nil
```

In order to return a good error, it's handling two edge cases. First, it cares whether it's going to say `1` or some number `%d`, so it can get the plural right. But the complexity is mostly from returning the right column. If begins by initializing the col number as if there are too few arguments, so it sets col after the last argument. Then it checks - if there are too many arguments, it sets col to the beginning of the first extra token instead.

The reason it "guesses" first, instead of using an if/else is because I prefer these 4 lines:
```
x := a
if something {
  x = b
}
```
compared to these 6 lines
```
var x <type>
if something {
  x = b
}  else {
  x = a
}
```


Reading your second comment more carefully, that may not have been your question.

The extra token, `mnemonic` is the token of the opcode itself. In an assembler, the "name" of an opcode is often called the "mnemonic" of the instruction.  I tried to keep using `name` for an argument that is a string, and mnemonic when it was a token. I needed to start passing the `mnemonic` token around so that I had access to the position (line and col) for good error messages.

You might wonder, "If you have the mnemonic token, don't you also have the name in `mnemoic.str`?"  Usually you do, but sometimes the `name` passed in here is "extended" to say something like "extract with 1 immediate" instead of just the name. (This happens when a pseudo spec is returned from `getSpec` with the extra info). So I pass both into `checkArgCount` so the error message can use that extended name, but point to the right spot on the line by using the mnemonic token.

Reply:
got it, thanks for explaining this method, and that sounds totally making sense to me!

Also thanks for the additional explanation on `name` and `mnemnoic.str` part, I was unaware of the pseudo ops part, but reporting error with `name` and computing locational info based on `mnemonic.str` seems reasonable.